mono-native 6.12.0.161 build error for Fedora 38 host

Not sure what causes this. Kernel 6.2.x? GLIBC? GCC 13?
```
... [C parts of Mono build up to this point]
CSC     [build-linux] gensources.exe
Makefile:20: warning: overriding recipe for target 'test-local'
../../build/library.make:208: warning: ignoring old recipe for target 'test-local'
Makefile:20: warning: overriding recipe for target 'test-local'
../../build/library.make:208: warning: ignoring old recipe for target 'test-local'

=================================================================
        Native Crash Reporting
=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
        Native stacktrace:
=================================================================
        0x4a3dd2 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x4a4179 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x455301 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x41ba1f - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x7fe2e73bcb70 - /lib64/libc.so.6 : 
        0x49fee8 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x45a67d - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x41d1a6 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x6143ef - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x616e8f - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x572d81 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x424c59 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x42af1c - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : mono_main
        0x41b1ff - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : 
        0x7fe2e73a6b4a - /lib64/libc.so.6 : 
        0x7fe2e73a6c0b - /lib64/libc.so.6 : __libc_start_main
        0x41b7d5 - /home/zozo/dtd-yocto-4.2/tmp-sicom-glibc/work/x86_64-linux/mono-native/6.12.0.161-r0/mono-6.12.0.122/mono/mini/mono : _start

=================================================================
        Telemetry Dumper:
=================================================================
No threads attached to runtime.

=================================================================
        External Debugger Dump:
=================================================================
mono_gdb_render_native_backtraces not supported on this platform, unable to find gdb or lldb

=================================================================
        Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x49fee8):0x49fed8  83 c2 01 48 83 fa 01 0f 87 a9 00 00 00 83 e8 05  ...H............
0x49fee8  c6 03 e8 89 43 01 4d 85 f6 74 4d 4c 89 6b 06 41  ....C.M..tML.k.A
0x49fef8  bc 0e 00 00 00 c6 43 05 08 83 fd 0e 75 4a 4d 85  ......C.....uJM.
0x49ff08  ff 74 03 41 89 2f 89 ee 48 89 df e8 58 c1 ff ff  .t.A./..H...X...
```
The  same crash occurs a couple of times for different DLLs until the build finally stops.

FWIW, I looked at the Fedora SRPM / spec for possible fixes.

- adding Fedora patches didn't work
- using `-O1` instead of `-O2` didn't work

I am out of ideas at the moment.

@ajlennon Any idea?

--- a/recipes-mono/mono/mono-6.12.0.161/disable-mmap-MAP_32BIT-support.patch
+++ b/recipes-mono/mono/mono-6.12.0.161/disable-mmap-MAP_32BIT-support.patch
@@ -1,0 +1,47 @@
+From e019d4ac4735e774d0a3c15bb36d5eb3ebfa3053 Mon Sep 17 00:00:00 2001
+From: Raphael Robatsch <raphael-git@tapesoftware.net>
+Date: Tue, 9 May 2023 11:55:02 +0200
+Subject: [PATCH] Disable mmap(MAP_32BIT) support
+
+mmap(2) with flag MAP_32BIT can erroneously return ENOMEM on recent
+kernels. Disable MAP_32BIT support for now.
+
+Reference: https://lore.kernel.org/linux-mm/cb8dc31a-fef2-1d09-f133-e9f7b9f9e77a@sony.com/
+Reference: https://lore.kernel.org/all/20230414185919.4175572-1-Liam.Howlett@oracle.com/T/#m00a0ac8a72bf2f26711b7f8cc56612a8ef62c3d0
+---
+ mono/mini/mini-amd64.h    | 2 --
+ mono/utils/mono-codeman.c | 4 ----
+ 2 files changed, 6 deletions(-)
+
+diff --git a/mono/mini/mini-amd64.h b/mono/mini/mini-amd64.h
+index b321743b67d..0a81bb4bd6b 100644
+--- a/mono/mini/mini-amd64.h
++++ b/mono/mini/mini-amd64.h
+@@ -390,9 +390,7 @@
+ 
+ #endif /* !HOST_WIN32 */
+ 
+-#if !defined(__linux__)
+ #define MONO_ARCH_NOMAP32BIT 1
+-#endif
+ 
+ #ifdef TARGET_WIN32
+ #define MONO_AMD64_ARG_REG1 AMD64_RCX
+diff --git a/mono/utils/mono-codeman.c b/mono/utils/mono-codeman.c
+index 234aac4b0ca..5eccda92bd0 100644
+--- a/mono/utils/mono-codeman.c
++++ b/mono/utils/mono-codeman.c
+@@ -68,11 +68,7 @@ static const MonoCodeManagerCallbacks *code_manager_callbacks;
+ #define MAX_WASTAGE 32
+ #define MIN_BSIZE 32
+
+-#ifdef __x86_64__
+-#define ARCH_MAP_FLAGS MONO_MMAP_32BIT
+-#else
+ #define ARCH_MAP_FLAGS 0
+-#endif
+
+ #define MONO_PROT_RWX (MONO_MMAP_READ|MONO_MMAP_WRITE|MONO_MMAP_EXEC|MONO_MMAP_JIT)
+
+--
+2.40.1

--- a/recipes-mono/mono/mono-native-6.xx-base.inc
+++ b/recipes-mono/mono/mono-native-6.xx-base.inc
@@ -7,7 +7,7 @@ RPROVIDES:${PN} += "nativesdk-mono"
 PROVIDES += "nativesdk-mono"
 
 do_compile() {
-    make EXTERNAL_MCS="${S}/mcs/class/lib/monolite/basic.exe" EXTERNAL_RUNTIME="${S}/foo/bar/mono"
+    oe_runmake EXTERNAL_MCS="${S}/mcs/class/lib/monolite/basic.exe" EXTERNAL_RUNTIME="${S}/foo/bar/mono"
 }
 
 EDEPENDS_X11 =+ "libgdiplus-native"

--- a/recipes-mono/mono/mono-native_6.12.0.161.bb
+++ b/recipes-mono/mono/mono-native_6.12.0.161.bb
@@ -27,6 +27,7 @@ SRC_URI = "http://download.mono-project.com/sources/mono/mono-${BASEPV}.tar.xz \
            file://0036-Revert-2020-02-Start-a-dedicated-thread-for-MERP-cra.patch \
            file://0037-Allow-nfloat-to-be-in-the-ObjCRuntime-namespace-and-.patch \
            file://0039-aot-Prepend-the-assembly-name-to-the-names-of-gshare.patch \
+           file://disable-mmap-MAP_32BIT-support.patch \
 "
 
 addtask fixup_config after do_patch before do_configure

--- a/recipes-mono/mono/mono_6.12.0.161.bb
+++ b/recipes-mono/mono/mono_6.12.0.161.bb
@@ -28,6 +28,7 @@ SRC_URI = "http://download.mono-project.com/sources/mono/mono-${BASEPV}.tar.xz \
            file://0036-Revert-2020-02-Start-a-dedicated-thread-for-MERP-cra.patch \
            file://0037-Allow-nfloat-to-be-in-the-ObjCRuntime-namespace-and-.patch \
            file://0039-aot-Prepend-the-assembly-name-to-the-names-of-gshare.patch \
+           file://disable-mmap-MAP_32BIT-support.patch \
 "
 
 

